### PR TITLE
Issue 99

### LINKED ISSUE #99: NoMethodError when running reports
This issue comes from #94 (Problem 8)

**System:** Fedora 23

### Current Behavior

NoMethodError when running reports

```
NoMethodError (undefined method `occurred_at' for nil:NilClass):
  app/models/shift.rb:69:in `block in breaks_duration'
  app/models/shift.rb:68:in `each'
  lib/csv_report_generator.rb:21:in `generate_report'
  app/reports/concerns/csv_generatable.rb:45:in `to_csv'
  app/controllers/hours_reports_controller.rb:11:in `block (2 levels) in show'
  app/controllers/hours_reports_controller.rb:10:in `show'
```

#### Steps to Reproduce

Suspect this happens when EndBreakDateTime is nil.

### Desired Behavior

Creating a report should return a reasonable default instead of blowing up with a `NoMethodError`.

### PR DESCRIPTION
**Pull request by crawfoal**
Resolve #99

If a volunteer forgets to sign back in from a break, detect this when generating the hours report, and include an error message in the corresponding cell in the shift duration column (telling them we couldn't calculate the duration because there was incomplete information).

When I was reviewing the code for `HoursReport`, I noticed that the `pull_join` method could be cleaned up a little, so I went ahead and did that too (I wrote a spec to make sure I didn't change behavior). If you guys don't want it in this pull request, I can remove it from this branch.

I also want to display a flash message telling the admin to take a closer look at the report if we detect that there is missing info, but I'm not exactly sure how to go about doing that (or if it's even possible). I just wanted to make it obvious to the admin that spreadsheet was incomplete so that they didn't accidentally miss the error message in the spreadsheet.

### CODE REVIEW

File: app/reports/hours_report.rb [hunk 1]
```diff
@@ -28,12 +28,11 @@ def csv_headers
   #
   # @return [ActiveRecord::Relation]
   def pull_join
-    # TODO: Have yet to get a Railsy query working here
     Shift
       .includes(:work_site, :volunteer, :shift_events)
+      .completed
```
Comment (leesharma) [inline app/reports/hours_report.rb]: Nice refactoring! :+1:

**Comment by bjmllr** [comment]
Thanks for doing this! I learned a few things from reviewing it. It was not intentional that this was left unresponded-to for this long.

> I also want to display a flash message telling the admin to take a closer look at the report if we detect that there is missing info, but I'm not exactly sure how to go about doing that (or if it's even possible).

Since the hours report is only presented in CSV we can't do a flash *in* the report, though we could repeat the message somewhere else in the CSV. Right now the data for a given week is small, so I suspect it's okay as is. Other options might be to put a notice in the weekly email, or arrange for a flash to be displayed on the dashboard when the CSV download link is clicked, or build an HTML-based hours report, similar to the signatures report, and include a flash there.

File: app/reports/hours_report.rb (hunk 1 above)
Comment (bjmllr) [inline app/reports/hours_report.rb]: Deleting TODO comments is the best. :tada:

File: spec/reports/hours_report_spec.rb [hunk 2]
```diff
@@ -17,25 +17,32 @@
   it 'generates CSV' do
     # Create a shift inside our weekly report period
     FactoryGirl.create(:shift, :full, day: begin_date + 1.day)
-    round_trip = CSV.parse(report.to_csv)
-    expect(round_trip.size).to be > 1
-    expect(round_trip.first.size)
+    rows = CSV.parse(report.to_csv)
```
Comment (crawfoal) [inline spec/reports/hours_report_spec.rb]: Sounds good, I added this in.

**Comment by crawfoal** [comment]
@bjmllr Ya, I was wanting to respond with both the CSV file, and some HTML/JS that would show a flash message (so the file would start downloading, but they would also see a message pop up on the page). When I started looking into it though, it started to sound like the wrong approach to take... or maybe like it wasn't even possible to include both the CSV download and the HTML/JS in the response together.

I'll just leave it this way for now, and something could be done later if more emphasis is needed.

**Pull request merged**